src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpMetricSerializer.cs

```diff
@@ -0,0 +1,352 @@
+        if (meterVersion != null)
+        {
+            writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.InstrumentationScope_Version, meterVersion);
+        }
+
+        if (meterTags != null)
+        {
+            // TODO: DON'T HAVE ANY TESTS THAT EVALUATE THIS
+            foreach (var tag in meterTags)
+            {
+                writePosition = WriteTag(buffer, writePosition, tag, ProtobufOtlpMetricFieldNumberConstants.InstrumentationScope_Attributes);
+            }
+        }
+
+        ProtobufSerializer.WriteReservedLength(buffer, instrumentationScopeLengthPosition, writePosition - (instrumentationScopeLengthPosition + ReserveSizeForLength));
+
+        for (int i = 0; i < metrics.Count; i++)
+        {
+            writePosition = WriteMetric(buffer, writePosition, metrics[i]);
+        }
+
+        return writePosition;
+    }
+
+    private static int WriteMetric(byte[] buffer, int writePosition, in Metric metric)
+    {
+        writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.ScopeMetrics_Metrics, ProtobufWireType.LEN);
+        int metricLengthPosition = writePosition;
+        writePosition += ReserveSizeForLength;
+
+        writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Metric_Name, metric.Name);
+
+        if (metric.Description != null)
+        {
+            writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Metric_Description, metric.Description);
+        }
+
+        if (metric.Unit != null)
+        {
+            writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Metric_Unit, metric.Unit);
+        }
+
+        switch (metric.MetricType)
+        {
+            case MetricType.LongSum:
+            case MetricType.LongSumNonMonotonic:
+                {
+                    writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Metric_Data_Sum, ProtobufWireType.LEN);
+                    int metricTypeLengthPosition = writePosition;
+                    writePosition += ReserveSizeForLength;
+
+                    writePosition = ProtobufSerializer.WriteBoolWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Sum_Is_Monotonic, metric.MetricType == MetricType.LongSum);
+                    writePosition = ProtobufSerializer.WriteEnumWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Sum_Aggregation_Temporality, metric.Temporality == AggregationTemporality.Cumulative ? 2 : 1);
+
+                    foreach (var metricPoint in metric.GetMetricPoints())
```

Comment:
All of these `foreach` loops over `GetMetricPoints()` should use ref:

```csharp
foreach (ref readonly var metricPoint in metric.GetMetricPoints())
```

Did this code come from GenevaExporter? We should fix that there too! 🤣

Reply:
This one is much harder to prove. It depends on what the compiler determines it needs to do with `metricPoint`. It may or may not decide to make a copy. But the idea is the `ref` _always_ avoids that.

Here is a sharplab: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6zQGYmpBkIYBvGg3dM7kiFwDmDAFlyAAoAmAwoZQAFCF4eXABBMCtcaAYBCOjY+IBKFzcPQu8/BlwOfAYAXgY6DWpCwoAzaBhsMAALBmCFbCh08MiwGLiMBjj+zKHsjFwcgob8+oWGsoqAamqMweGeFgBxcNlygBkfX2CcuuWGAF95huIAdlLyq487pfd74i8zwNJQgMsiNEslVKk+ltgblFtdiv5VlUam8Fs1YG1OsFYI0GOiACY+SQATwYPUhQKmIzGXAm22ms3uhVcn2uL3WmwpOww+0OJzOFxRDQ+1yebMFN0KjK+LJ+pQiHDAozCky592Z1zEkR6GBgDASvl8sF82D4PgAKkSxLrsAaLVbBYVNbxtbrlXSRgA1bCSDgwWQYaA23W4LjYMS4doQDBen0wB0ee4AbWVkbxBnEkkBGFT6bEkgA8mJTVxcCx9Ybwc6YEZJHE4r4cgBdb52fGEknwhgHDBHfCnPwXNVS9y8HHBbO8Us23x23UAQmq5aNJt45stgn7vl7Rkg+HrABEYJIMNgGAAyM8MCdT23rhgLvUG5fF2csTfbri7+tCcocSQrhQYDmFkmWHBprxYM12igCAAHcADkoyOMQxGgHU8TdZRZwAUQQFQi1XLhglDAQIEacd2knHkez5PwckuMCPjA0UIJDMMIyjGNfTLXBN3Fa4wKTFMIDTDMsxzDNC2LUtEJrOs6ObFknRdXFWgJLhiSYFAGCgmCEKQjgULQmAMIpHC8JgAifGCZhGAyVN4OwARgNAkCPGzPSGC4GBYIYRCaKMqB0Nw/Di2CAASAAiZx7JExyBAlScvKjF5AvQhg0SvSjcFpZQr3XFhIoYlkJWWe5hQTahE39KAFQwY5sCJCAOAwYIGqalqAGk4jxFhcLzZReAwJsaBGGRQ0kOVasVQJOWmLi/QDKAgxodUPETAAxXgjzxfNGkaXBwmCOgRplX4SgSXizjqJMtp2vaDqOk7FMKWUCQ4YBpD1XB92az64xoUrpUlWg7BgLhykfCtjRfddVvud9yh3UQDyPE80ARs4Py/Pwf3wP8AJgQH7lsVTsHUzTcHlGbMMpeIkhSaB4aUrUTV1NsNJJWmuUTRtcrAXAuWsFnnTZsmKZJEZef5oWzQgKIYJSQVlLFjnNM7E8oF8cIRG4DAbpZMaoAm2aVXpBnwWgLN3R4aWoQFoW0GpDA7bmkE5YViAUidjXem1jBdZ4FyPDWhYIPtwX6SRCOhbneN3HDt34g9xXwWjpOZhTr3wTjsCIM1/3A9GaoC515qeHFGgWwYbCIYEZalq7cJa/KGQTSt4OFkqAA+LyfIoqiY/pJ3E7N935dT3BcBH7KWFLgPy+Gg3XrsKnptGFv6/bqAhzc9wVZ1cX21Nm2Xb5oeQXjhgD/ZtTj6l8+M9wLOlbAm+j85hhfa1su9av9/Ow9RgAgZeywjYm03m3Ja1toRn35pHEETsH4y3pC/cEPs/jzyLp3BoodWSj0GAg+I6dJhEJmFfQoBDYHPwntnKeJDT40M9q/Pe4FZ5YMXkiDhf8wKUNnkAhASI4DkCvkxVhso1YkmxB/TS3NpiOA4FAWAPBeEeB7qpHEVCHb0kTFo2WtCUi6P4VwPEwDGyNlAQsWUwAIAQEmgECAgF4LANajg646i1hrAggIhgAAeLKVFuEV3KoDEGNwgA

In there I call `metricPoint.GetSumLong()` to get the value which is analogous to what we do for counters in the code here.

The method generated NOT using `ref` is longer vs the `ref` one. Not an expert in x64 assembly but my guess is there is a copy being made in the longer one somewhere.

The general idea is `MetricPoint` is a BIG struct so you want to use it by `ref` to avoid having to make copies.